src/include/mpir_pset.h

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#ifndef MPIR_PSET_H_INCLUDED
+#define MPIR_PSET_H_INCLUDED
+
+#include "mpiimpl.h"
+
+/**
+ * @brief Process set structure
+ */
+struct MPIR_Pset {
+    char *uri;                  /* Name of the process set */
+    int size;                   /* Number of processes in the set */
+    bool is_valid;              /* True if process set is valid, false if not */
+    int *members;               /* Array of members of the process set (ranks) sorted in ascending order
+                                 * TODO: This should be optimized using compressed methods of process
+                                 * representation to improve scalability */
+};
+
+/**
+ * @brief Process set array structure based on UT_array data structure + reference counter and mutex
+ */
+struct MPIR_Pset_array {
+    UT_array *parray;           /* Array of psets */
+    MPID_Thread_mutex_t *mutex; /* Mutex to protect parray from concurrent accesses */
+};
+
+/**
+ * @brief   Get pset from pset array based on its name.
+ *
+ * @param   parray      Pset array in which the pset shall be found
+ * @param   pset_name   Name of the pset
+ * @param   pset        Returns the found pset
+ * @return  int         MPI_SUCCESS or MPI_ERR_OTHER
+ */
+int MPIR_Pset_by_name(MPIR_Pset_array * pset_array, const char *pset_name, MPIR_Pset ** pset);
+
+/**
+ * @brief   Get pset from pset array based on its index.
+ *
+ * @param   parray  Pset array in which the pset shall be found
+ * @param   idx     Index of the pset
+ * @param   pset    Returns the found pset
+ * @return  int     MPI_SUCCESS or MPI_ERR_OTHER
+ */
+int MPIR_Pset_by_idx(MPIR_Pset_array * pset_array, int idx, MPIR_Pset ** pset);
+
+/**
+ * @brief Initialize a pset array - must be called from within a lock
+ *
+ * @param pset_array    Pset array to initialize
+ * @param mutex         Optional pointer to mutex (mutex created if not NULL)
+ */
+void MPIR_Pset_array_init(MPIR_Pset_array ** pset_array, MPID_Thread_mutex_t * mutex);
+
+/**
+ * @brief Destroy a pset array - must be called from within a lock
+ *
+ * @param parray    Pset array to be destroyed
+ */
+void MPIR_Pset_array_destroy(MPIR_Pset_array * pset_array);
+
+/**
+ * @brief   Add a pset to a pset array.
+ *
+ * @param   parray  Pset array to which the pset shall be added
+ * @param   pset    Pset to be added
+ * @return  int     MPI_SUCCESS or MPI_ERR_OTHER
+ */
+int MPIR_Pset_array_add(MPIR_Pset_array * pset_array, MPIR_Pset * pset);
+
+/**
+ * @brief   Invalidate a pset in pset array.
+ *
+ * @param   parray      Pset array in which the pset shall be invalidated
+ * @param   pset_name   Name of the pset
+ * @return  int         MPI_SUCCESS or MPI_ERR_OTHER (not found)
+ */
+int MPIR_Pset_array_invalidate(MPIR_Pset_array * pset_array, char *pset_name);
```

Comment:
Since we only maintain a single global pset array, we don't need `pset_array` in the parameter.

Reply:
Right. Removed the parameter.